cabal-install/Distribution/Client/CmdClean.hs

```diff
@@ -0,0 +1,104 @@
+{-# LANGUAGE RecordWildCards #-}
+module Distribution.Client.CmdClean (cleanCommand, cleanAction) where
+
+import Prelude ()
+import Distribution.Client.Compat.Prelude
+
+import Distribution.Client.DistDirLayout
+    ( DistDirLayout(..), defaultDistDirLayout )
+import Distribution.Client.ProjectConfig
+    ( findProjectRoot )
+import Distribution.Client.Setup 
+    ( GlobalFlags )
+import Distribution.ReadE ( succeedReadE )
+import Distribution.Simple.Setup
+    ( Flag(..), toFlag, fromFlagOrDefault, flagToList, flagToMaybe
+    , optionDistPref, optionVerbosity, falseArg
+    )
+import Distribution.Simple.Command 
+    ( CommandUI(..), option, reqArg )
+import Distribution.Simple.Utils 
+    ( info, wrapText )
+import Distribution.Verbosity 
+    ( Verbosity, normal )
+
+import Control.Exception 
+    ( throwIO )
+import System.Directory
+    ( removeDirectoryRecursive, doesDirectoryExist )
+
+data CleanFlags = CleanFlags 
+    { cleanSaveConfig :: Flag Bool
+    , cleanVerbosity :: Flag Verbosity
+    , cleanDistDir :: Flag FilePath
+    , cleanProjectFile :: Flag FilePath
+    } deriving (Eq)
+
+defaultCleanFlags :: CleanFlags
+defaultCleanFlags = CleanFlags 
+    { cleanSaveConfig = toFlag False
+    , cleanVerbosity = toFlag normal
+    , cleanDistDir = NoFlag
+    , cleanProjectFile = mempty
+    }
+
+cleanCommand :: CommandUI CleanFlags
+cleanCommand = CommandUI
+    { commandName = "new-clean"
+    , commandSynopsis = "Clean the package store and remove temporary files."
+    , commandUsage = \pname ->
+        "Usage: " ++ pname ++ " new-clean [FLAGS]\n"
+    , commandDescription  = Just $ \_ -> wrapText $
+        "Removes all temporary files created during the building process "
+     ++ "(.hi, .o, preprocessed sources, etc.) and also empties out the "
+     ++ "local caches (by default).\n\n"
+    , commandNotes = Nothing
+    , commandDefaultFlags = defaultCleanFlags
+    , commandOptions = \showOrParseArgs -> 
+        [ optionVerbosity
+            cleanVerbosity (\v flags -> flags { cleanVerbosity = v })
+        , optionDistPref
+            cleanDistDir (\dd flags -> flags { cleanDistDir = dd })
+            showOrParseArgs
+        , option [] ["project-file"]
+            "Set the name of the cabal.project file to search for in parent directories"
+            cleanProjectFile (\pf flags -> flags {cleanProjectFile = pf})
+            (reqArg "FILE" (succeedReadE Flag) flagToList)
+        , option ['s'] ["save-configure"]
+            "Save configuration, only remove build artefacts"
+            cleanSaveConfig (\sc flags -> flags { cleanSaveConfig = sc })
+            falseArg
+        ]
+  }
+
+cleanAction :: CleanFlags -> [String] -> GlobalFlags -> IO ()
+cleanAction CleanFlags{..} _ _ = do
```

Comment:
Maybe fail on unexpected arguments, so that, if we want to extend `new-clean`, we won't be breaking any code mistakenly relying on being able to pass junk on the command line.

Reply:
You're right, that's pretty important/basic.